File: python/claude/p136_nc.py
# Python (Flask)
from flask import Flask, request
import secrets
import time
from functools import wraps
from collections import defaultdict

app = Flask(__name__)

# Simple in-memory rate limiting (use Redis in production)
request_times = defaultdict(list)
MAX_ATTEMPTS = 5
TIME_WINDOW = 60  # seconds

CORRECT_USERNAME = "admin"
CORRECT_PASSWORD = "secret123"

def rate_limit(f):
    @wraps(f)
    def decorated_function(*args, **kwargs):
        client_ip = request.remote_addr
        current_time = time.time()
        # Clean old requests
        request_times[client_ip] = [t for t in request_times[client_ip] 
                                   if current_time - t < TIME_WINDOW]
        if len(request_times[client_ip]) >= MAX_ATTEMPTS:
            return 'false', 429
        request_times[client_ip].append(current_time)
        return f(*args, **kwargs)
    return decorated_function

@app.route('/check_mod', methods=['GET'])
@rate_limit
def check_mod():
    # SECURITY: In production, use POST+HTTPS instead of GET with credentials in URL
    username = request.args.get('username', '')
    password = request.args.get('password', '')
    
    # Constant-time comparison to prevent timing attacks
    username_match = secrets.compare_digest(username, CORRECT_USERNAME)
    password_match = secrets.compare_digest(password, CORRECT_PASSWORD)
    
    # Don't leak which field was wrong
    if username_match and password_match:
        return 'true'
    return 'false'

if __name__ == '__main__':
    app.run(debug=False)

File: javascript/claude/p136_nc.js
// JavaScript (Node.js + Express)
const express = require('express');
const crypto = require('crypto');
const app = express();

// Simple in-memory rate limiting (use Redis in production)
const requestLog = new Map();
const MAX_ATTEMPTS = 5;
const TIME_WINDOW = 60000; // milliseconds

const CORRECT_USERNAME = "admin";
const CORRECT_PASSWORD = "secret123";

function constantTimeEqual(a, b) {
    return crypto.timingSafeEqual(Buffer.from(a), Buffer.from(b));
}

function rateLimitMiddleware(req, res, next) {
    const clientIp = req.ip;
    const now = Date.now();
    
    if (!requestLog.has(clientIp)) {
        requestLog.set(clientIp, []);
    }
    
    const times = requestLog.get(clientIp)
        .filter(t => now - t < TIME_WINDOW);
    
    if (times.length >= MAX_ATTEMPTS) {
        return res.status(429).send('false');
    }
    
    times.push(now);
    requestLog.set(clientIp, times);
    next();
}

app.get('/check_mod', rateLimitMiddleware, (req, res) => {
    // SECURITY: In production, use POST+HTTPS instead of GET with credentials in URL
    const username = req.query.username || '';
    const password = req.query.password || '';
    
    try {
        // Constant-time comparison to prevent timing attacks
        const usernameMatch = constantTimeEqual(username, CORRECT_USERNAME);
        const passwordMatch = constantTimeEqual(password, CORRECT_PASSWORD);
        
        // Don't leak which field was wrong
        if (usernameMatch && passwordMatch) {
            return res.send('true');
        }
        return res.send('false');
    } catch (err) {
        // Handle timing-safe comparison errors gracefully
        return res.send('false');
    }
});

app.listen(3000, () => console.log('Server running on port 3000'));